### PR TITLE
Multi-process support

### PR DESCRIPTION
**Pull request by qianxichen233**
This PR adds multi-process support (clone syscall, execve syscall, exit syscall) for lind within wasmtime.

### CODE REVIEW

**Comment by qianxichen233** [comment]
Implemented features and related files:
- moved lind_syscall interface from wasi_preview1 to a standalone module `lind-common` and add cageid parameter to lind_syscall. Related files are `crates/lind-common/src/lib.rs`
- Implemented clone syscall and exit syscall. Related files are `crates/lind-multi-process/src/clone_constants.rs`, `crates/lind-multi-process/src/lib.rs`, `crates/wasmtime/src/runtime/externals.rs`, `crates/wasmtime/src/runtime/func.rs`, `crates/wasmtime/src/runtime/func/typed.rs`, `crates/wasmtime/src/runtime/store.rs`
- Created a global cage counter to make sure the process will only exit after all cages were exited (instead of main process exiting). Related files are `crates/lind-utils/src/lib.rs`
- Implemented execve syscall. Related files are `crates/lind-multi-process/src/lib.rs`, `src/commands/run.rs`

**Review by rennergade — CHANGES_REQUESTED** [review]
Unfortunately its in our best interests to split this PR as well.

Realistically it should be

PR1: the changes needed to interface with the base rustposix PR and the base glibc PR. This should go to main.

PR2: the changes for multi-processing.

**Comment by qianxichen233** [comment]
> Unfortunately its in our best interests to split this PR as well.
> 
> Realistically it should be
> 
> PR1: the changes needed to interface with the base rustposix PR and the base glibc PR. This should go to main.
> 
> PR2: the changes for multi-processing.

Splitted. This PR now contains mainly multi-process stuff only

File: crates/lind-multi-process/src/clone_constants.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+pub const CLONE_NEWCGROUP: u64 =    0x02000000;	/* New cgroup namespace.  */
+pub const CLONE_NEWUTS: u64 =	0x04000000;	/* New utsname group.  */
+pub const CLONE_NEWIPC: u64 =	0x08000000;	/* New ipcs.  */
+pub const CLONE_NEWUSER: u64 =	0x10000000;	/* New user namespace.  */
+pub const CLONE_NEWPID: u64 =	0x20000000;	/* New pid namespace.  */
+pub const CLONE_NEWNET: u64 =	0x40000000;	/* New network namespace.  */
+pub const CLONE_IO: u64 =	0x80000000;	/* Clone I/O context.  */
+/* cloning flags intersect with CSIGNAL so can be used only with unshare and
+   clone3 syscalls.  */
+pub const CLONE_NEWTIME: u64 =	0x00000080;      /* New time namespace */
```
Comment (rennergade) [inline crates/lind-multi-process/src/clone_constants.rs:45]: newline

File: crates/lind-multi-process/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,1259 @@
...
+        let get_cx = self.get_cx.clone();
+        let parent_pid = self.pid;
+
+        // set up unwind callback function
+        let store = caller.as_context_mut().0;
+        let is_parent_thread = store.is_thread();
+        store.set_on_called(Box::new(move |mut store| {
+            // let unwind_stack_finish;
+
+            // let address = cloned_address as *mut u64;
```
Comment (rennergade) [inline crates/lind-multi-process/src/lib.rs:293]: remove commented code?

File: crates/lind-multi-process/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,1259 @@
+#![allow(dead_code)]
+
+use anyhow::{anyhow, Result};
+use rawposix::safeposix::dispatcher::lind_syscall_api;
+use wasi_common::WasiCtx;
+use wasmtime_lind_utils::{parse_env_var, LindCageManager};
+
+use std::ffi::CStr;
```
Comment (rennergade) [inline crates/lind-multi-process/src/lib.rs:8]: this file needs a ton more detail for each function in terms of what they do

File: crates/wasmtime/src/runtime/func.rs [hunk 4]
```diff
@@ -1059,15 +1059,43 @@ impl Func {
...
-            let func_ref = func_ref.as_ref();
-            (func_ref.array_call)(
-                func_ref.vmctx,
-                caller.cast::<VMOpaqueContext>(),
-                params_and_returns,
-                params_and_returns_capacity,
-            )
-        })
+        let mut result;
+
```
Comment (rennergade) [inline crates/wasmtime/src/runtime/func.rs:1063]: needs comments explaining what this does

File: crates/wasmtime/src/runtime/func/typed.rs [hunk 5]
```diff
@@ -211,16 +211,38 @@ where
...
-            (func_ref.array_call)(
-                func_ref.vmctx,
-                VMOpaqueContext::from_vmcontext(caller),
-                (storage as *mut Storage<_, _>) as *mut ValRaw,
-                mem::size_of_val::<Storage<_, _>>(storage) / mem::size_of::<ValRaw>(),
-            );
-        });
+        let mut result;
+
+        loop {
```
Comment (rennergade) [inline crates/wasmtime/src/runtime/func/typed.rs:216]: comments

File: crates/wasmtime/src/runtime/instance.rs [hunk 6]
```diff
@@ -452,6 +454,31 @@ impl Instance {
         self._get_export(store, entity, export_name_index)
     }
 
+    pub fn get_stack_pointer(&self, mut store: impl AsContextMut) -> Result<i32, ()> {
+        if let Some(sp_extern) = self.get_export(store.as_context_mut(), "__stack_pointer") {
```
Comment (rennergade) [inline crates/wasmtime/src/runtime/instance.rs:458]: does this have to be nested like this?

File: src/commands/run.rs [hunk 7]
```diff
@@ -232,6 +259,125 @@ impl RunCommand {
         Ok(())
     }
 
+    fn execute_with_lind(mut self, lind_manager: Arc<LindCageManager>, pid: i32, next_cageid: Arc<AtomicU64>) -> Result<Vec<Val>> {
```
Comment (rennergade) [inline src/commands/run.rs:262]: would appreciate more comments in this function

File: crates/wasmtime/src/runtime/instance.rs (hunk 6 above)
Comment (qianxichen233) [reply crates/wasmtime/src/runtime/instance.rs:458]: Updated, there is one unnecessary Err(()). Except this, all other Err(()) has its own error implication.

File: Cargo.lock
Comment (Yaxuan-w) [inline Cargo.lock:1]: Do we need to include this file in repo? this should be auto-generated right?

**Review by Yaxuan-w — COMMENTED** [review]
great work!

File: Cargo.lock
Comment (rennergade) [reply Cargo.lock:1]: yeah, i dont think so? @qianxichen233 take a look at this

File: crates/lind-multi-process/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,1323 @@
...
+
+    // shared-memory version of fork syscall, used to create a new thread
+    // This is very similar to normal fork syscall, except the memory is not copied
+    // and the saved unwind context need to be carefully copied and managed since parent
+    // and child are operating two copies to unwind data in the same memory
+    // Function Argument:
+    // * stack_addr: child's base stack address
+    // * stack_size: child's stack size
+    // * child_tid: the address of the child's thread id. This should be set by wasmtime
+    pub fn fork_shared_call(&self, mut caller: &mut Caller<'_, T>,
```
Comment (rennergade) [inline crates/lind-multi-process/src/lib.rs:523]: Is it possible to rename this function? If this is to create new threads and not "cages" it not really a fork right?

File: crates/lind-utils/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    if let Some(pos) = env_var.find('=') {
+        // If '=' is found, return the key and value as String and Some(String)
+        let key = env_var[..pos].to_string();
+        let value = env_var[pos + 1..].to_string();
+        (key, Some(value))
+    } else {
+        // If '=' is not found, return the whole string as the key and None for the value
+        (env_var.to_string(), None)
+    }
+}
```
Comment (rennergade) [inline crates/lind-utils/src/lib.rs:58]: newline

File: crates/wasmtime/src/lib.rs [hunk 10]
```diff
@@ -271,7 +271,7 @@
 //!
 //! [manifest]: https://github.com/bytecodealliance/wasmtime/blob/main/crates/wasmtime/Cargo.toml
 
-#![deny(missing_docs)]
+// #![deny(missing_docs)]
```
Comment (rennergade) [inline crates/wasmtime/src/lib.rs:274]: remove instead of comment?

**Pull request merged**